lib/migration_priv.h

```diff
@@ -33,94 +33,95 @@
 
 #include <linux/vfio.h>
 
-/*
- * FSM to simplify saving device state.
- */
-enum migr_iter_state {
-    VFIO_USER_MIGR_ITER_STATE_INITIAL,
-    VFIO_USER_MIGR_ITER_STATE_STARTED,
-    VFIO_USER_MIGR_ITER_STATE_DATA_PREPARED,
-    VFIO_USER_MIGR_ITER_STATE_FINISHED
-};
-
 struct migration {
-    /*
-     * TODO if the user provides an FD then should mmap it and use the migration
-     * registers in the file
-     */
-    struct vfio_user_migration_info info;
+    uint64_t flags;
+    enum vfio_user_device_mig_state state;
     size_t pgsize;
     vfu_migration_callbacks_t callbacks;
-    uint64_t data_offset;
-
-    /*
-     * This is only for the saving state. The resuming state is simpler so we
-     * don't need it.
-     */
-    struct {
-        enum migr_iter_state state;
-        uint64_t pending_bytes;
-        uint64_t offset;
-        uint64_t size;
-    } iter;
 };
 
-struct migr_state_data {
-    uint32_t state;
-    const char *name;
+/*
+ * valid migration state transitions
+ * each number corresponds to a FROM state and each bit a TO state
+ * if the bit is set then the transition is allowed
+ * the indices of each state are those in the vfio_user_device_mig_state enum
+ */
+static const char transitions[8] = {
```

Comment:
you probably want to assert max nr of states.

Reply:
also you don't want a variable defined in the header like this.

Reply:
Where should all these be defined then, `migration.c`?

Reply:
yes